File: src/http/request.rs
use std::str;
use std::collections::HashMap;
use regex;

#[derive(Debug)]
pub struct Request
{
	pub method: Option<String>,
	pub protocol: Option<String>,
	pub host: Option<String>,
	pub port: Option<String>,
	pub uri: Option<String>,
	pub version: Option<String>,
	pub headers: HashMap<String, String>,
	pub cookies: HashMap<String, String>,
}

impl Request
{
	pub fn new(raw_request: &[u8]) -> Request {
		let mut eol_index = 0;
		for ch in raw_request {
			eol_index += 1;
			if *ch == 10u8 {
				break;
			}
		}

		let string_re = regex::Regex::new(r"^(\S+) (?:(https?|tcp|ftp|ssh)://)?([\-\.a-zA-Z0-9]+)(?::(\d{1,5}))?((?:/[\-\.a-zA-Z0-9\?\&=]*)*)? (\S+)").unwrap();

		let mut request_info = Request {
			method: None,
			protocol: None,
			host: None,
			port: None,
			uri: None,
			version: None,
			headers: HashMap::new(),
			cookies: HashMap::new(),
		};

		if let Some(re_search_results) = string_re.captures(str::from_utf8(&raw_request[0 .. eol_index]).unwrap()) {
			if let Some(group) = re_search_results.get(1) {
				request_info.method = Some(String::from(group.as_str()));
			}
			if let Some(group) = re_search_results.get(2) {
				request_info.protocol = Some(String::from(group.as_str()));
			}
			if let Some(group) = re_search_results.get(3) {
				request_info.host = Some(String::from(group.as_str()));
			}
			if let Some(group) = re_search_results.get(4) {
				request_info.port = Some(String::from(group.as_str()));
			}
			else {
				match request_info.protocol.clone().unwrap().as_str() {
					p if p == "https" => {
						request_info.port = Some(String::from("443"));
					},
					p if p == "ftp" => {
						request_info.port = Some(String::from("21"));
					},
					p if p == "ssh" => {
						request_info.port = Some(String::from("22"));
					},
					_ => {
						request_info.port = Some(String::from("80"));
					},
				}
			}
			if let Some(group) = re_search_results.get(5) {
				request_info.uri = Some(String::from(group.as_str()));
			}
			if let Some(group) = re_search_results.get(6) {
				request_info.version = Some(String::from(group.as_str()));
			}
		}

		return request_info;
	}
}

File: src/http/request.cpp
#include "request.hpp"

#include <regex>
#include <sstream>

namespace http
{
	Request::Request(std::vector<uint8_t> raw_request) {
		std::regex request_re(R"(^(\S+) (?:(https?|tcp|ftp|ssh)://)?([\-\.a-zA-Z0-9]+)(?::(\d{1,5}))?((?:/[\-\.a-zA-Z0-9\?\&=]*)*)? (\S+))");
		std::regex headers_re(R"(^(.*?): ?(.*)\s)");

		std::stringstream request_stream(reinterpret_cast<char*>(raw_request.data()));

		std::string line;
		std::smatch regex_result;
		std::getline(request_stream, line);
		if (std::regex_search(line, regex_result, request_re)) {
			if (regex_result[1].matched) {
				method = regex_result[1];
			}
			if (regex_result[2].matched) {
				protocol = regex_result[2];
			}
			if (regex_result[3].matched) {
				host = regex_result[3];
			}
			if (regex_result[4].matched) {
				std::stringstream(regex_result[4]) >> port;
			}
			else {
				if (protocol == "https") {
					port = 443;
				}
				else if (protocol == "ftp") {
					port = 21;
				}
				else if (protocol == "ssh") {
					port = 22;
				}
				else
					port = 80;
			}
			if (regex_result[5].matched) {
				uri = regex_result[5];
			}
			if (regex_result[6].matched) {
				version = regex_result[6];
			}
		}

		while (std::getline(request_stream, line))
		{
			if (std::regex_search(line, regex_result, headers_re)) {
				headers.emplace(regex_result[1], regex_result[2]);
			}
		}
	}
	std::string Request::create_request() {
		std::stringstream request;

		request << method << ' ' << uri << ' ' << version << "\r\n";
		for (const auto & header : headers) {
			request << header.first << ": " << header.second << "\r\n";
		}

		request << "\r\n";

		return request.str();
	}
}

